### ISSUE TITLE
[Bug] Video thread crash in Intel HD 5500 mobile

### ISSUE THREAD

**Issue by mmillar-bolis** [andrei-drexler/ironwail#102]
Hey there! I am experiencing a bug where only the video dies during a segment of gameplay in Ironwail.

What happens is I can start a completely vanilla game of Quake, play for a bit, and eventually the video thread will crash while the rest of the game process is technically functional. Whatever causes this to happen varies by level, however `start.bsp` is probably the easiest one to reproduce the issue.

On the `start` map, once I pass through a difficulty portal, moving into any of the four episode hallways causes the crash, but not until then. Loading campaign maps directly can get past this, but each one has it's own spots that trigger the crash, such as `e1m1` by merely walking forward towards the first door on the left, or `e1m3` a bit after going through the gold key door and dropping down into the water. And because the latter map's crash point is so late, it's possible to watch through the demo file without actually triggering the issue.

What happens when the video crashes is that the viewport is stuck on the last drawn frame (unless one alt-tabs out which causes it to turn black), while sound, input, AI, and everything else continues to function. If one can remember what position the menu cursor was last on, it's actually possible to exit the game elegantly.

Some details about my affected system:

| Key | Description |
|----:|:------------|
| Hardware | Panasonic Toughbook CF-31 v5 |
| CPU | Intel Core i5 5300U |
| GPU | Intel HD Graphics 5500 |
| OS | Windows 10 1809 x64 |
| GPU Driver | Intel 20.19.15.5058 (Panasonic) |

(FYI: I am using the latest version of the Intel driver that is available from my system's manufacturer, and Intel's latest generic installer flags itself as incompatible otherwise.)

Aside from some mouse and keyboard bindings, as well as toggling on developer mode for logging, I have not changed any other settings. I have also tested this issue in Ironwail versions 0.5.0 & 0.6.0 as well as commits `17ff33e` and `7d3d9df` and the crash remains consistent. Nothing illuminating appears to be written to the log when this happens, but I figured I would attach them for completeness sake.

[ironwail-0.5.0-console.log](https://github.com/andrei-drexler/ironwail/files/9042245/ironwail-0.5.0-console.log)
[ironwail-0.6.0-console.log](https://github.com/andrei-drexler/ironwail/files/9042246/ironwail-0.6.0-console.log)
[ironwail-17ff33e-console.log](https://github.com/andrei-drexler/ironwail/files/9042248/ironwail-17ff33e-console.log)
[ironwail-7d3d9df-console.log](https://github.com/andrei-drexler/ironwail/files/9042247/ironwail-7d3d9df-console.log)
[ironwail.cfg](https://github.com/andrei-drexler/ironwail/files/9042266/ironwail.cfg.txt)

Is there any way I could provide more information on this issue?

**Comment by andrei-drexler** [comment]
Hey! One thing you could try is running the game with `-gldebug`, that *might* provide additional info (besides a couple of shader recompilation warnings at startup that I already know about). I actually have a tablet with a GPU using basically the same architecture (Gen 8LP vs Gen8), though, and the driver isn't reporting any errors. but my driver is even older than yours (20.19.15.4835), so who knows... There are also a couple of switches that disable some OpenGL extensions (`-nobufferstorage`, `-nomultibind`, `-noclipcontrol`), which might help isolate the issue.

**Comment by mmillar-bolis** [comment]
These were great ideas, thank you so much! I sat down and tried each extra option. Like you pointed out, `-gldebug` only added one extra, unrelated warning about shader recompiling, but I will be keeping this in mind for future use.

As for the other switches (`-nobufferstorage`, `-nomultibind`, `-noclipcontrol`), sadly no combination seems to alleviate the thread crash on my system.

**Comment by andrei-drexler** [comment]
Unfortunately, I kind of expected this, but it was worth a try.

You mentioned testing a couple of recent commits - I've created a new branch ([issue-102](https://github.com/andrei-drexler/ironwail/tree/issue-102)) for various experiments that might help, could you please give it a go and see if anything changes? It creates a debug context (which might provide additional info compared to just enabling debug output on a regular context) and requests robust access (in case there are out-of-bounds buffer reads/writes). `-gldebug` is no longer necessary, it's effectively always on.

**Comment by mmillar-bolis** [comment]
Wow, I really appreciate that!

I'm going to be honest with you, I am a complete novice at this, so I apologize ahead of time if I'm not too helpful yet. I went ahead and built the code how I usually do, using Msys2 for Windows. I wasn't sure whether to build a debug executable in light of what you said, so I did so anyway (using `DEBUG=1`).

[Ironwail64-Issue_102-1bb5b2e.log](https://github.com/andrei-drexler/ironwail/files/9075443/Ironwail64-Issue_102-1bb5b2e.log)

I ran the executable through `gdb`, though again, I must admit that I don't yet know what I'm doing and didn't receive much additional output either way, as you can see from the game log above. Unfortunately, the crash still happens in the identical places as standard builds.

Could performing process minidumps both pre and post crash provide you with any better information? Or do I perhaps need to set up Visual Studio to get some more robust debugging (system symbols and whatnot)?

**Comment by andrei-drexler** [comment]
Sorry about the delay, very little to no energy left for this project during the past few weeks. Also, thanks for all the info - if you're not a driver engineer working for Intel or Panasonic I'd say you're already helping more than anyone could reasonably be expected to in this situation, so don't worry about it!

Unfortunately, I'm pretty sure this is a driver bug, and honestly I think the chances of coming up with a solution are pretty slim. Compiling with Visual Studio might help rule out any differences introduced by MinGW, but I wouldn't expect the debugger to help much. Since the main thread carries on as if nothing happened, the crash must be in a worker thread, but it will likely occur long after the underlying condition is triggered by the main thread - that is, if we're even dealing with a crash and not just a silent error state. And if it is indeed a crash, and if we can actually get a callstack at the right moment, due to Panasonic's customizations I'm not even sure Intel's driver symbols will match. This might still work, but it's a long shot.

Another option is a black box approach - applying various changes to see if there's any difference in output and maybe eventually isolate the root cause. That can take a really long time, though, and the chances of success are also quite low. That being said, I have pushed a couple more commits to the issue branch - could you please try the latest code to see if anything changes?

**Comment by mmillar-bolis** [comment]
Hey, sorry for the late response. Yes, I went ahead and built against the latest revision. Unfortunately, no change; same stall.

What you are saying makes perfect sense and while I'm committed to continuing to test, I totally understand if it makes more sense to close this issue for now, given how slim the chances are of tracking the cause down.

What's important is that I made you aware of a minor, fringe bug, in case it comes up again later. Your engine fork still works wonderfully on my Nvidia desktop, so thanks very much! If it's easier to manage, you could kill the `issue-102` branch and I can keep testing on the main one and just drop you a comment if things ever change.

In either event, I don't mind continuing to build and test, but I understand that this is going to be a low priority due to how niche it is.

**Comment by marcovirtual** [comment]
Hello everybody!

I'm having the EXACT same problem reported by @mmillar-bolis. The crashes happen at almost the same spots he reported.

Device name	DESKTOP-B0A2V0M
Processor	Intel(R) Core(TM) i5-5200U CPU @ 2.20GHz   2.20 GHz
Installed RAM	4,00 GB (3,90 GB usable)
System type	64-bit operating system, x64-based processor
Edition	Windows 10 Home
Version	22H2
Installed on	‎26/‎04/‎2022
Intel HD Graphics 5500

PS: I love Ironwail.
